compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1494,7 +1498,10 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             TypingMode::Coherence => false,
             // Avoid using the global cache when we're defining opaque types
             // as their hidden type may impact the result of candidate selection.
-            TypingMode::Analysis { defining_opaque_types } => defining_opaque_types.is_empty(),
+            // HACK: hack
+            TypingMode::Analysis { defining_opaque_types } => {
+                defining_opaque_types.is_empty() || !pred.has_opaque_types()
```

Comment:
given that we use `can_use_global_caches` both on access and on write, could we change this to instead check that `pred` does not contain an opaque from `defining_opaque_types`, i.e. it only contains rigid opaque types

Reply:
I actually tested that and it doesn't improve perf very much lol

Reply:
yeah perf only goes from 4:07 to 4:04 when i implement a simple visitor to check if it contains an opaque in the defines list

Reply:
> yeah perf only goes from 4:07 to 4:04 when i implement a simple visitor to check if it contains an opaque in the defines list

that seems weird to me :thinking: what's causing the exponential blowup then, reuse between different MIR borrowck queries?